File: bukkit/freezer/src/main/java/dialight/freezer/FreezerTool.java
package dialight.freezer;

import dialight.compatibility.PlayerInventoryBc;
import dialight.extensions.*;
import dialight.maingui.MainGuiTool;
import dialight.misc.ActionInvoker;
import dialight.misc.Colorizer;
import dialight.misc.ItemStackBuilder;
import dialight.offlinelib.OfflineEntity;
import dialight.misc.player.UuidPlayer;
import dialight.toollib.Tool;
import dialight.toollib.events.ToolInteractEvent;
import org.bukkit.Material;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.PluginDescriptionFile;

public class FreezerTool extends Tool {

    public static final String ID = "freezer";
    private final Freezer proj;
    private final ItemStack item;

    public FreezerTool(Freezer proj) {
        super(ID);
        this.proj = proj;
        PluginDescriptionFile desc = proj.getPlugin().getDescription();
        this.item = new ItemStackBuilder(Material.ICE)
                .displayName(Colorizer.apply("|a|Замораживатель игроков"))
                .lore(Colorizer.asList(
                        "|w|Выбор игроков",
                        "|a|ПКМ по игроку*|y|: заморозить/разморозить игрока",
                        "|y|*Игрока можно пометить на расстоянии.",
                        "|a|ПКМ|y|: открыть замораживатель",
//                        "|a|Shift|y|+|a|ПКМ|y|: добавить замороженных",
//                        "|y| в список телепортируемых",
                        "",
                        "|g|Плагин: |y|Замораживатель",
                        "|g|Версия: |y|v" + desc.getVersion()
                ))
                .let(this::applyId)
                .build();
    }

    private void toggleFreeze(ActionInvoker invoker, UuidPlayer target) {
        Frozen frozen = proj.get(target);
        if(frozen == null) {
            proj.register(new Frozen(target, target.getLocation(), invoker, "freezer tool action"));
        } else {
            proj.unregister(invoker, target);
        }
    }

    @Override
    public void onClick(ToolInteractEvent e) {
        switch (e.getAction()) {
            case LEFT_CLICK: if(!e.isSneaking()) {

            } else {

            } break;
            case RIGHT_CLICK: if(!e.isSneaking()) {
                ActionInvoker invoker = new ActionInvoker(proj.getOfflineLib().getUuidPlayer(e.getPlayer()));
                if (e.getTarget() == ToolInteractEvent.Target.ENTITY) {
                    ToolInteractEvent.Entity eevent = (ToolInteractEvent.Entity) e;
                    if (eevent.getEntity().getType() == EntityType.PLAYER) {
                        Player target = (Player) eevent.getEntity();
                        toggleFreeze(invoker, proj.getOfflineLib().getUuidPlayer(target));
                    }
                } else {
                    Entity p2 = LocationEx.of(e.getPlayer().getEyeLocation()).getEnByDirection(20.0, 1.5, EntityType.PLAYER, OfflineEntity.TYPE);
                    if (p2 == null) {
                        proj.getGuilib().openGui(e.getPlayer(), proj.getGui());
                    } else {
                        if(p2.getType() == EntityType.PLAYER) {
                            Player target = (Player) p2;
                            toggleFreeze(invoker, proj.getOfflineLib().getUuidPlayer(target));
                        } else if(p2.getType() == OfflineEntity.TYPE) {
                            UuidPlayer up = proj.getOfflineLib().getUuidPlayerExByEntity(p2);
                            if(up != null) {
                                toggleFreeze(invoker, up);
                            }
                        }
                    }
                }
            } else {

            } break;
            case DROP: {
                if(proj.getMaingui() != null) {
                    MainGuiTool tool = proj.getToollib().get(MainGuiTool.class);
                    if(tool != null) {
                        PlayerInventoryBc.of(e.getPlayer().getInventory()).setItemInMainHand(tool.createItem());
                    }
                }
                proj.getGuilib().clearStory(e.getPlayer());
            } break;
        }
    }

    @Override public ItemStack createItem() {
        return item;
    }

}

File: sponge/freezer/src/main/kotlin/dialight/freezer/FreezerTool.kt
package dialight.freezer

import dialight.maingui.MainGuiTool
import dialight.extensions.Utils
import dialight.teleporter.Teleporter
import dialight.toollib.Tool
import dialight.toollib.events.ToolInteractEvent
import jekarus.colorizer.Text_colorized
import jekarus.colorizer.Text_colorizedList
import org.spongepowered.api.entity.EntityTypes
import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.item.ItemTypes

class FreezerTool(val plugin: FreezerPlugin) : Tool(FreezerTool.ID) {

    companion object {
        val ID = "freezer"
    }

    override val type = ItemTypes.ICE
    override val title = Text_colorized("|a|Замораживатель игроков")
    override val lore = Text_colorizedList(
        "|w|Заморозка",
        "|a|ЛКМ по игроку|y|: заморозить/разморозить игрока",
        "|w|Выбор игроков",
        "|a|ПКМ|y|: открыть замораживатель",
        "|a|Shift|y|+|a|ПКМ|y|: добавить замороженных",
        "|y| в список телепортируемых",
        "",
        "|g|Плагин: |y|Замораживатель",
        "|g|Версия: |y|v" + plugin.container.version.orElse("null")
    )

    override fun onClick(e: ToolInteractEvent) {
        when (e.action) {
            ToolInteractEvent.Action.LEFT_CLICK -> if (!e.sneaking) {
                if (e.type == ToolInteractEvent.Type.ENTITY) {
                    e as ToolInteractEvent.Entity
                    if (e.entity.type == EntityTypes.PLAYER) {
                        plugin.freezer.invoke(e.player, Freezer.Action.TOGGLE, e.entity as Player)
                    }
                } else {
                    val p2 = Utils.getEnByDirection(e.player, 20.0, 1.5, EntityTypes.PLAYER) as? Player
                    if (p2 != null) {
                        plugin.freezer.invoke(e.player, Freezer.Action.TOGGLE, p2)
                    }
                }
            } else {

            }
            ToolInteractEvent.Action.RIGHT_CLICK -> if (!e.sneaking) {
                plugin.guilib?.also { guilib ->
                    guilib.openGui(e.player, plugin.freezergui)
                }
            } else {
                val teleporter = plugin.teleporter.teleporter
                teleporter.invoke(e.player, Teleporter.Action.UNTAG, Teleporter.Group.ALL)
                val result = Teleporter.Result()
                plugin.freezer.forEachOnline { uuid, name ->
                    result.add(teleporter.invoke(e.player, Teleporter.Action.TAG, uuid, name))
                }
                result.sendReport(e.player)
            }
            ToolInteractEvent.Action.DROP -> {
                plugin.eh?.also { eh ->
                    plugin.toollib.giveTool(e.player, MainGuiTool.ID)
                }
                plugin.guilib?.clearStory(e.player)
            }
        }
    }

}
